### PR TITLE
[fix](DECIMALV3)  Fix the error in DECIMALV3 when explicitly casting.

### PR DESCRIPTION
**Pull request by Mryange**
# Proposed changes

before
```
mysql [test]>select cast(1 as DECIMALV3(16, 2)) /  cast(3 as DECIMALV3(16, 2));
+-----------------------------------------------------------+
| CAST(1 AS DECIMALV3(16, 2)) / CAST(3 AS DECIMALV3(16, 2)) |
+-----------------------------------------------------------+
|                                                      0.00 |
+-----------------------------------------------------------+


mysql [test]>select * from divtest;
+------+------+
| id   | val  |
+------+------+
|    3 | 5.00 |
|    2 | 4.00 |
|    1 | 3.00 |
+------+------+

mysql [test]>select cast(1 as decimalv3(16,2)) / val from divtest;
+-------------------------------------+
| CAST(1 AS DECIMALV3(16, 2)) / `val` |
+-------------------------------------+
|                                   0 |
|                                   0 |
|                                   0 |
+-------------------------------------+
```

after
```
mysql [test]>select cast(1 as DECIMALV3(16, 2)) /  cast(3 as DECIMALV3(16, 2));
+-----------------------------------------------------------+
| CAST(1 AS DECIMALV3(16, 2)) / CAST(3 AS DECIMALV3(16, 2)) |
+-----------------------------------------------------------+
|                                                      0.33 |
+-----------------------------------------------------------+

mysql [test]>select cast(1 as decimalv3(16,2)) / val from divtest;
+-------------------------------------+
| CAST(1 AS DECIMALV3(16, 2)) / `val` |
+-------------------------------------+
|                            0.250000 |
|                            0.200000 |
|                            0.333333 |
+-------------------------------------+
```
This is because in the previous code, the constant 1.000 would be transformed into 1.

remove "ReduceType

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Mryange** [comment]
run buildall

**Review by HappenLee — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/analysis/Expr.java [hunk 1]
```diff
@@ -284,6 +284,11 @@ public boolean apply(Expr arg) {
     // Needed for properly capturing expr precedences in the SQL string.
     protected boolean printSqlInParens = false;
 
+    // The code is used to distinguish whether an expression is explicitly cast or
+    // not. For example, different treatments should be applied to a computed value
+    // of 1.000 and a value of 1.000 obtained through casting.
+    protected boolean isCasted = false;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/Expr.java:290]: should set back to default value in `reset()` function

**Comment by morrySnow** [comment]
btw, the divide result should be decimal(18, 2), but current result's scale is 6

File: fe/fe-core/src/main/java/org/apache/doris/analysis/Expr.java (hunk 1 above)
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/Expr.java:290]: should set back to default value in `reset()` function

**Review by morrySnow — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java [hunk 2]
```diff
@@ -536,7 +536,7 @@ private void analyzeDecimalV3Op(Type t1, Type t2) throws AnalysisException {
     public void analyzeImpl(Analyzer analyzer) throws AnalysisException {
         if (VectorizedUtil.isVectorized()) {
             for (Expr child : children) {
-                if (child instanceof DecimalLiteral && child.getType().isDecimalV3()) {
+                if (child instanceof DecimalLiteral && child.getType().isDecimalV3() && !child.getIsCasted()) {
                     ((DecimalLiteral) child).tryToReduceType();
```
Comment (starocean999) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java:540]: try sql `select 1.0000000 / cast(3 as DECIMALV3(1, 0));` seems this pr can't handle such case, IMO,`tryToReduceType()` method is ambiguous, it's not clear why need this function, is it possible to remove it?

Comment (Mryange) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java:540]: I asked the author about this code, and they said it was meant to reduce precision during calculations.  now it seems better to remove it.

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run p1

**Review by morrySnow — APPROVED** [review]

**Pull request merged**